src/wallet/wallet2.cpp

```diff
@@ -590,11 +607,49 @@ static uint64_t decodeRct(const rct::rctSig & rv, const crypto::public_key pub,
   }
 }
 //----------------------------------------------------------------------------------------------------
-bool wallet2::wallet_generate_key_image_helper(const cryptonote::account_keys& ack, const crypto::public_key& tx_public_key, size_t real_output_index, cryptonote::keypair& in_ephemeral, crypto::key_image& ki)
+static uint64_t decodeRct_onetime(const rct::rctSig & rv, const crypto::public_key pub, const crypto::secret_key &sec, uint64_t received_mode, const crypto::secret_key* onetime_h, unsigned int i, rct::key & mask)
 {
-  if (!cryptonote::generate_key_image_helper(ack, tx_public_key, real_output_index, in_ephemeral, ki))
-    return false;
-  return true;
+  if (received_mode == 2)
+  {
+    crypto::secret_key sec2;
+    sc_mul((unsigned char*)&sec2, (const unsigned char*)&sec, (const unsigned char*)onetime_h);
+    return decodeRct(rv, pub, sec2, i, mask);
+  }
+  else
+  {
+    return decodeRct(rv, pub, sec, i, mask);
+  }
+}
+//----------------------------------------------------------------------------------------------------
+struct payment_id_info {
+  tx_extra_nonce extra_nonce;
+  bool has_payment_id32 = false;
+  bool has_payment_id8  = false;
+  crypto::hash  payment_id32 = null_hash ;
+  crypto::hash8 payment_id8  = null_hash8;
+  crypto::secret_key* onetime_h = nullptr;
+  crypto::secret_key  onetime_h_real;
+};
+static void get_payment_id_info(const std::vector<tx_extra_field>& tx_extra_fields, const crypto::secret_key& viewkey, payment_id_info& pID_info) {
+  if (find_tx_extra_field_by_type(tx_extra_fields, pID_info.extra_nonce))
+  {
+    pID_info.has_payment_id32 = get_payment_id_from_tx_extra_nonce          (pID_info.extra_nonce.nonce, pID_info.payment_id32);
+    pID_info.has_payment_id8  = get_encrypted_payment_id_from_tx_extra_nonce(pID_info.extra_nonce.nonce, pID_info.payment_id8 );
+    // if 8-byte payment ID is found, copy it as-is to 32-byte payment ID regardless of whether it's encrypted or not
+    if (pID_info.has_payment_id8)
+    {
+      memcpy(&pID_info.payment_id32, &pID_info.payment_id8, 8);
+      memset(&pID_info.payment_id32 + 8, 0, 24);
+    }
+    // try to get h for onetime address scheme
+    char data[64];
+    memcpy(data     , &viewkey              , 32);
+    memcpy(data + 32, &pID_info.payment_id32, 32);
+    crypto::hash_to_scalar(data, 2 * HASH_SIZE, reinterpret_cast<crypto::ec_scalar&>(pID_info.onetime_h_real));
+    // check if the following equation holds: (H(viewkey, pID) - pID) mod 256 == 0
+    if (pID_info.payment_id32.data[0] == pID_info.onetime_h_real.data[0])
+      pID_info.onetime_h = &pID_info.onetime_h_real;
```

Comment:
I think 0.4% of non-onetime payment ids will also trigger this. This will just cause the onetime test to be done, but be otherwise fine, right ?